sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeExpressions.scala

```diff
@@ -1963,3 +1963,64 @@ case class Epoch(child: Expression, timeZoneId: Option[String] = None)
     defineCodeGen(ctx, ev, c => s"$dtu.getEpoch($c, $zid)")
   }
 }
+
+@ExpressionDescription(
+  usage = "_FUNC_(field, source) - Extracts a part of the date/timestamp.",
+  arguments = """
+    Arguments:
+      * field - selects which part of the source should be extracted. Supported string values are:
+                ["MILLENNIUM", "CENTURY", "DECADE", "YEAR", "QUARTER", "MONTH",
+                 "WEEK", "DAY", "DAYOFWEEK", "DOW", "ISODOW", "DOY",
+                 "HOUR", "MINUTE", "SECOND"]
+      * source - a date (or timestamp) column from where `field` should be extracted
+  """,
+  examples = """
+    Examples:
+      > SELECT _FUNC_('YEAR', TIMESTAMP '2019-08-12 01:00:00.123456');
+       2019
+      > SELECT _FUNC_('week', timestamp'2019-08-12 01:00:00.123456');
+       33
+      > SELECT _FUNC_('doy', DATE'2019-08-12');
+       224
+  """,
+  since = "3.0.0")
+case class DatePart(field: Expression, source: Expression, child: Expression)
+  extends RuntimeReplaceable {
+
+  def this(field: Expression, source: Expression) {
+    this(field, source, {
+      if (!field.foldable) {
+        throw new AnalysisException("The field parameter needs to be a foldable string value.")
```

Comment:
This follows the standard? It seems pgSQL accept a non-foldable value;
```
postgres=# create table t (a text, b timestamp);
CREATE TABLE
postgres=# insert into t values ('year', '2019-08-12 01:00:00.123456');
INSERT 0 1
postgres=# select * from t;
  a   |             b              
------+----------------------------
 year | 2019-08-12 01:00:00.123456
(1 row)

postgres=# select date_part(a, b) from t;
 date_part 
-----------
      2019
(1 row)
```

Reply:
According to PostgreSQL docs https://www.postgresql.org/docs/11/functions-datetime.html#FUNCTIONS-DATETIME-EXTRACT:

>    _source_ must be a value expression ...  the _**field**_ parameter needs to be **a string value**

Accepting _field_ as an expression is undocumented feature. We could support that separately if it is needed.

Reply:
I checked the mysql behaivour and it doesn't support the case. So, its ok to keep as it is.